### PR TITLE
[Merged by Bors] - feat: lint `attribute [instance] instName in`

### PR DESCRIPTION
**Pull request by grunweg**
Contrary to what users would expect, this defines a global instance.

This linter would have caught #13144 and #13170.
[Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/ConcreteCategory.2EinstFunLike.20became.20a.20.20global.20instance)

Co-authored by: @adomani

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by adomani — COMMENTED** [review]
This looks good!

I do not know my way around `attribute` syntax, but is the alternative implementation below have some drawbacks?  It seems simpler to me.

Also, could you add a `#guard_msgs`ed test?

```lean
/-- Whether a syntax element is adding an `instance` attribute without a `local` modifier. -/
def is_attribute_instance_in : Syntax → Bool
  | `(command|attribute [instance] $_decl:ident in $_) => true
  | `(command|attribute [instance $_priority] $_decl:ident in $_) => true
  | _ => false

def attributeInstanceIn : Linter where run := withSetOptionIn fun stx => do
  unless getLinterAttributeInstanceIn (← getOptions) do
    return
  if (← MonadState.get).messages.hasErrors then
    return
  if is_attribute_instance_in stx then
    Linter.logLint linter.attributeInstanceIn stx[0] m!
      "careful: `attribute [instance] ... in` declarations are surprising: \
      they are **not** limited to the subsequent declaration, \
      but define global instances \
      please remove the `in` or make this a `local instance` instead"
```

File: Mathlib/Tactic/Linter/AttributeInstanceIn.lean [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      -- "in" the next declaration (the case we're linting against) depends on the parent.
+
+      -- Its parent determines whether we have a scoped instance, and whether it's "in".
+      -- scoped instance, no in: parent is the instance declaration of positive index
+      -- non-scoped, no in: parent is empty
+      -- "in" attribute: something non-trivial of index 0
+      match chain.get? 0 with
+      -- This is the case we're interested in
+      | some (_next, 0) => Linter.logLint linter.attributeInstanceIn head m!
+      "careful: `attribute [instance] ... in` declarations are surprising:
```
Comment (adomani) [inline Mathlib/Tactic/Linter/AttributeInstanceIn.lean:71]: I do not have an opinion on this, but in `Mathlib` `m!` is never separated from its string.

File: Mathlib/Tactic/Linter/AttributeInstanceIn.lean [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      they are **not** limited to the subsequent declaration,
+      but define global instances
```
[the remark this answers is not in the document]
Comment (grunweg) [reply Mathlib/Tactic/Linter/AttributeInstanceIn.lean]: I had not, and was puzzled about the indentation. Thanks, this solves it!

File: Mathlib/Tactic/Linter/AttributeInstanceIn.lean (hunk 1 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/AttributeInstanceIn.lean]: The whole string is way too long for the line, but the `m!` must start on the same line as the `logLint` call... I've tried something in between now.

**Comment by grunweg** [comment]
I wrote a test yesterday (of course!), but forgot to commit it :facepalm: I've updated it to use your example instance now (and added you as an author of the tests), since this avoid mathlib imports.

File: Mathlib/Tactic/Linter/AttributeInstanceIn.lean [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+def attributeInstanceIn : Linter where run := withSetOptionIn fun stx => do
+  unless getLinterAttributeInstanceIn (← getOptions) do
+    return
+  if (← MonadState.get).messages.hasErrors then
+    return
+  match stx.findStack? (fun _ ↦ true) is_attribute_instance_in with
+  | some ((head, 0)::chain) =>
+    -- An instance attribute is always the first child of its parent syntax.
+    if is_attribute_instance_in head then
```
[the remark this answers is not in the document]
Comment (grunweg) [reply Mathlib/Tactic/Linter/AttributeInstanceIn.lean]: Oh, yes, this is also correct.

**Comment by grunweg** [comment]
I just tried the logic you proposed. It works, so I switched to it (and added you an an author).
I tweaked one test to consider nested syntax (which worked), so I guess `attribute`s are always the first syntax.

File: Mathlib/Tactic/Linter/AttributeInstanceIn.lean [hunk 4]
```diff
@@ -0,0 +1,71 @@
+/-
+Copyright (c) 2024 Michael Rothgang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Rothgang, Damiano Testa
+-/
+
+import Lean.Elab.Command
```
Comment (adomani) [inline Mathlib/Tactic/Linter/AttributeInstanceIn.lean:7]:
```suggestion
import Lean.Elab.Command
import Lean.Linter.Util
```
I do not know if you were tricked by `Shake`: this import is needed for `logLint`.

**Review by adomani — APPROVED** [review]
I feel like I have had a bit of influence in this PR, so I am going to approve, but will not maintainer merge.

With respect to merging this linter into the "generic" one: I think that this can happen at a later stage, if/when the other PR is more ready.  At the moment, I still consider it experimental.

Thanks!

File: Mathlib/Tactic/Linter/AttributeInstanceIn.lean (hunk 4 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/AttributeInstanceIn.lean:7]: I was not tricked by shake - I removed this import manually and realised later that it wouldn't build...

**Comment by grunweg** [comment]
!bench

File: Mathlib/Tactic/Linter/AttributeInstanceIn.lean [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+# Linter for `attribute instance in` declarations
+
+The syntax `attribute [instance] instName in` can be used to accidentally create a global instance.
+This is **not** obvious from reading the code, and in fact happened twice during the port,
+hence, we lint against it.
+
+*Example*: before this was discovered, `Mathlib/Topology/Category/TopCat/Basic.lean`
+contained the following code:
+```
+-- Porting note: cannot find a coercion to function otherwise
```
Comment (semorrison) [inline Mathlib/Tactic/Linter/AttributeInstanceIn.lean:20]:
```suggestion
```

File: Mathlib/Tactic/Linter/AttributeInstanceIn.lean [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+Instead, remove the `in` or make this a `local instance` instead.
+-/
+def attributeInstanceIn : Linter where run := withSetOptionIn fun stx => do
+  unless getLinterAttributeInstanceIn (← getOptions) do
+    return
+  if (← MonadState.get).messages.hasErrors then
+    return
+  if is_attribute_instance_in stx then
+    Linter.logLint linter.attributeInstanceIn stx[0] m!
+      "careful: `attribute [instance] ... in` declarations are surprising:\n\
```
Comment (semorrison) [inline Mathlib/Tactic/Linter/AttributeInstanceIn.lean:64]: Either remove careful, or change to some word already used to introduce linter explanations?

**Comment by semorrison** [comment]
Two minor requests, otherwise

bors d+

File: Mathlib/Tactic/Linter/AttributeInstanceIn.lean (hunk 6 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/AttributeInstanceIn.lean:64]: Fair point: this linter is not meant to be advisory, but a hard error - I have removed the "careful".
```suggestion
      "`attribute [instance] ... in` declarations are surprising:\n\

**Comment by grunweg** [comment]
Thanks for the fast review.
bors r+

File: Mathlib/Tactic/Linter/AttributeInstanceIn.lean [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+namespace attributeInstanceLinter
+
+/-- Gets the value of the `linter.attributeInstanceIn` option. -/
+def getLinterAttributeInstanceIn (o : Options) : Bool :=
+  Linter.getLinterValue linter.attributeInstanceIn o
+
+/-- Whether a syntax element is adding an `instance` attribute without a `local` modifier. -/
+def is_attribute_instance_in (stx : Syntax) : Bool :=
+  match stx with
+  | `(command|attribute [instance] $_decl:ident in $_) => true
```
Comment (eric-wieser) [inline Mathlib/Tactic/Linter/AttributeInstanceIn.lean:47]: This should handle the case with multiple idents too

Comment (eric-wieser) [reply Mathlib/Tactic/Linter/AttributeInstanceIn.lean:47]: In fact, it should probably also accept a list of attributes that _contains_ `instance`

**Pull request closed without merging**